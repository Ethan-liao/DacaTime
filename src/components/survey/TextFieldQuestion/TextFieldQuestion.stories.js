import React from 'react';
import { storiesOf, action, linkTo } from '@kadira/storybook';
import { component as TextFieldQuestion } from './index';
import { loremIpsum } from '../../../stories/util';

storiesOf('Text field question', module).add('name input', () =>
  <TextFieldQuestion
    questionId={1}
    name="name"
    prompt="What is your name?"
    fields={[
      { label: 'First name', name: 'fname' },
      { label: 'Middle name', name: 'mname' },
      { label: 'Last name', name: 'lname' },
    ]}
    explanation="Last name is also known as family name."
    dispatch={action('updated-field')}
  />,
);
