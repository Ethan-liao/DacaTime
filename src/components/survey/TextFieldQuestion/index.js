import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import R from 'ramda';
import QuestionContainer from '../QuestionContainer';
import { updateField, updateError } from '../../../actions';
import { typeAndPattern } from '../../../util';
import * as DataLists from '../../basic/PlaceLists';

const changeHandler = dispatch => targetKeyPath => event => {
  const value = event.target.value;
  dispatch(updateField(targetKeyPath, value));
};

const TextFieldQuestion = props => {
  const blurHandler = (name, field) => event => {
    event.target.classList.add('attempted');
    if (!event.target.checkValidity() || event.target.value === '') {
      props.dispatch(updateError(name, field, true));
    } else if (props.errors[name]) {
      props.dispatch(updateError(name, field, false));
    }
  };

  const createTextField = ({ label, name }, idx) => {
    const updateKeyPath = [props.name, name];
    const value = R.pathOr('', ['value', name], props);
    const { type, pattern, list, errorMessage } = typeAndPattern(
      props.validators[idx],
    );
    return (
      <div key={idx}>
        {idx === 0 || <br />}

        <p className="control-label">
          {label}
        </p>
        <input
          type={type}
          className="survey-input"
          key={name}
          placeholder={label}
          value={value}
          onChange={changeHandler(props.dispatch)(updateKeyPath)}
          onBlur={blurHandler(props.name, name)}
          list={list}
          pattern={pattern}
        />
        {props.errors[props.name] &&
          props.errors[props.name][name] &&
          <p className="error-message">
            {errorMessage}
          </p>}
        {list &&
          <datalist id={list}>
            {DataLists[props.validators[idx]]}
          </datalist>}
      </div>
    );
  };

  const fields = R.addIndex(R.map)(createTextField)(props.fields);

  return (
    <QuestionContainer {...props}>
      {fields}
    </QuestionContainer>
  );
};
TextFieldQuestion.propTypes = {
  questionId: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  prompt: PropTypes.string.isRequired,
  explanation: PropTypes.string,
  fields: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
    }),
  ).isRequired,
  dispatch: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  dispatch,
});

const mapStateToProps = (state, ownProps) => ({
  value: state.form[ownProps.name],
  errors: state.global.errors,
  ...ownProps,
});

export const component = TextFieldQuestion;
export default connect(mapStateToProps, mapDispatchToProps)(component);
