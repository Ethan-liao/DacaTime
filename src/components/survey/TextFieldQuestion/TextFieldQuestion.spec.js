import React from 'react';
import { combineReducers } from 'redux';
import TextFieldQuestion from './index';
import { textQuestion } from '../../../util/testData';
import reducer from '../../../reducers';
import { setup } from '../../../util/testUtils';
import renderer from 'react-test-renderer';

const { component } = setup(TextFieldQuestion, reducer, textQuestion);

it('renders without crashing', () => {
  renderer.create(component);
});

it('renders correctly', () => {
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
