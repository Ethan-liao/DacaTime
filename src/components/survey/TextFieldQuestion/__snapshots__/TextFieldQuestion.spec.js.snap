exports[`test renders correctly 1`] = `
<div
  className="dt-panel dt-question-container">
  <div
    className="dt-question-header">
    When does your DACA status expire?
  </div>
  <div>
    <p
      className="control-label">
      MM/DD/YYYY
    </p>
    <input
      className="survey-input"
      list={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      pattern="^\\d{2}\\/\\d{2}\\/\\d{4}$"
      placeholder="MM/DD/YYYY"
      type="text"
      value="" />
  </div>
</div>
`;
